app/jobs/batch_autoscaling.py

```diff
@@ -0,0 +1,77 @@
+import boto3
+
+def get_compute_environment_capacity(compute_environment_name, region):
+    batch_client = boto3.client('batch', region_name=region)
+    response = batch_client.describe_compute_environments(computeEnvironments=[compute_environment_name])
+    compute_resources = response['computeEnvironments'][0]['computeResources']
+    return {'minvCpus': compute_resources['minvCpus'], 'maxvCpus': compute_resources['maxvCpus']}
+
+
+def set_compute_environment_min_capacity(vcpu_min_capacity, compute_environment_name, region):
+    batch_client = boto3.client('batch', region_name=region)
+    batch_client.update_compute_environment(computeEnvironment=compute_environment_name, computeResources={'minvCpus': vcpu_min_capacity})
+
+
+def check_pending_jobs_counts(queue_name, region, job_status_list=('RUNNABLE', 'STARTING', 'RUNNING')):
+    return {job_status: get_jobs_count(queue_name, region, job_status) for job_status in job_status_list}
+
+
+def autoscale_compute_environments(batch_configurations, dry_run=False):
+    results = []
+    for bc in batch_configurations:
+        try:
+            process_results = process_batch_configuration(bc, dry_run)
+            results.append({'batch_configuration': bc, 'process_results': process_results})
+        except Exception as e:
+            error_type = str(type(e))
+            error_args = e.args
+            results.append({'batch_configuration': bc, "error_type": error_type, "error_args": error_args})
+    return results
+
+
+def process_batch_configuration(batch_configuration, dry_run=False):
+    job_vcpus, queue_name, region = (batch_configuration[i] for i in ('vcpus', 'queue_name', 'region'))
+    compute_environment_name = find_compute_environment_name(queue_name, region)
+    current_capacity = get_compute_environment_capacity(compute_environment_name, region)
+    current_vcpu_min, current_vcpu_max = (current_capacity[i] for i in ('minvCpus', 'maxvCpus'))
+
+    pending_jobs_counts = check_pending_jobs_counts(queue_name, region)
+    total_jobs_count = sum(pending_jobs_counts.values())
```

Comment:
If we are doing a SUM here, rather then len(set union);  you want to be sure that you query the states in reverse order:  RUNNING, then STARTING, then RUNNABLE.   This will ensure you do not double-count jobs that have advanced from one state to the next as you do your separate requests.

Reply:
Hmm, that's a good suggestion, I'll do this change

Reply:
This reminds me of something important we haven't yet had a chance to look into.

It used to be that the lowmem queue would run only alignment jobs.   When that was the case, it was easy to limit the max number of alignment jobs running concurrently using the max vcpus on the compute env.   If we are now using that queue and env to also run other types of jobs, it may be impossible to use the max vcpu param to only control the max number of alignment jobs.   It's nice to have that sort of control, because of the way they compete for chunks.

A similar problem occurs in the highmem queue that is shared by host filering, postprocessing, etc.

I think for both problems we should actually create separate queues for host filtering and alignment.   The host filtering queue would share the high mem compute env, and the alignment queue would share the lowmem compute env.  However, the priority of the host filtering queue would be higher than the priority of the himem queue, and similarly the priority of the alignment queue would be higher than the priority of the lowmme queue.  That way alignment and host filtering jobs would be given priority over other jobs and stages;  this would ensure better utilization of the rapsearch and gsnap tiers.

Reply:
However, if we did that, we would lose the independent control for max number of alignment jobs.   Perhaps, then, alignment would need its own separate queue AND separate compute environment.

Reply:
Well, I would need to change the logic if we wanted to share the same compute environment for two different queues. This piece of code is assuming that each queue is connected to a single compute environment. Is that a correct assumption?